### ISSUE TITLE
ZFS deduplication produce excessive cpu usage by SHA256

### ISSUE THREAD

**Issue by pavel-odintsov** [zfsonlinux/zfs#2887]
Hello, folks!

I tried to copy 356Mb partition on pool with enabled deduplication (zfs set dedup=off data):

``` bash
for i in `seq 1 100`; do zfs send data/debian-7.0-x86_64-minimal@111114 |zfs recv data/$i; done
```

And when I look at top I was really amazed by cpu consumption by kernel:

``` bash
Cpu(s):  0.1%us, !!!42.6%sy!!!,  0.0%ni, 44.6%id, 12.7%wa,  0.0%hi,  0.1%si,  0.0%st
```

Detailed kernel performance top:

``` bash
Samples: 144K of event 'cycles', Event count (approx.): 33829255512
 48,26%  [kernel]            [k] SHA256Transform
  3,43%  [kernel]            [k] fletcher_4_incremental_native
  2,63%  [kernel]            [k] mutex_lock
  2,37%  [kernel]            [k] _spin_lock_irqsave
  2,33%  [kernel]            [k] mutex_unlock
  2,05%  [kernel]            [k] _spin_unlock_irqrestore
  1,47%  [kernel]            [k] lzjb_compress
  1,25%  [kernel]            [k] memmove
  1,16%  [kernel]            [k] copy_user_generic_string
  0,92%  [kernel]            [k] memset
  0,85%  [kernel]            [k] kmem_alloc_debug
  0,85%  [kernel]            [k] spl_kmem_cache_alloc
  0,77%  [kernel]            [k] native_write_msr_safe
  0,75%  [kernel]            [k] kfree
  0,75%  [kernel]            [k] schedule
  0,71%  [kernel]            [k] _spin_trylock
  0,67%  [kernel]            [k] __kmalloc
  0,65%  [kernel]            [k] memcpy
```

Maybe you can try to integrate Intel QuickAssist technology with hardware acceleration of hashing and compression: https://01.org/packet-processing/intel%C2%AE-quickassist-technology-drivers-and-patches ?

Intel Quick Assist can provide complete hardware acceleration of zlib and sha-256 calculation: http://www.intel.ru/content/dam/www/public/us/en/documents/product-briefs/quickassist-adapter-8920-brief.pdf

And my company (FastVPS Eesti OU) can provide Quick Assist chip with test server  if anybody interested in integration of this feature. 

ZFS: mirror 2 x SSHD drive ST1000DX001-1CM162
Platform: Intel(R) Xeon(R) CPU E5-2603 v2 @ 1.80GHz
CentOS 6.

**Comment by pavel-odintsov** [comment]
OK, we have troubles with Crypto API.

But current version of deduplication in ZFS is completely unusable because this issue. I tried to copy data from volume with dedup=off to volume with dedup=on and ZFS eaten whole my CPU on very slow ZFS mirror from 6 SATA drives!

But my cpu is not an i486 from 1989 and I'm using modern server with Intel Xeon E5-1607 v2 @ 3.00GHz.

And my htop screen looks like this:
```bash
  1  [||||||||||||||||||||||||||||||||||||||||||||||||94.7%]     Tasks: 28, 4 thr; 6 running
  2  [||||||                                          10.0%]     Load average: 4.18 5.15 5.81 
  3  [||||||||||||||||||||||||||||||||||||||||||||||||94.1%]     Uptime: 4 days, 05:05:12
  4  [||||||||||||||||||||||||||||||||||||||||||||||||94.0%]
  Mem[||||||||||||||||||||||||||||||||||||||||||6318/7799MB]
  Swp[|                                           5/20479MB]
```

This load produced only by ZFS and I haven't any running services there.  CPU eaten by SHA256Transform function as mentioned in first post. 

Maybe you could increase priority of this feature in roadmap? Because this issue makes "deduplication" feature on Linux completely absent.

**Comment by kpande** [comment]
seems a bit arbitrary for Linus to have decided to limit the Cryptographic API to GPL-only applications. Seems counterintuitive for a "free software" to have restrictions on such usage. Aha. The old "BSD is truly free" and "GPL forces you to feel free" thingy.

**Comment by pavel-odintsov** [comment]
Maybe we can try to use something like this http://cryptodev-linux.org/index.html and attach to this API via standard interfaces of cryptodev like user-space application?

**Comment by pdxChen** [comment]
I suspect that SHA256Transform is not the optimized SHA256 in the kernel crypto library, but a pretty slow software function that zfs called.  ZFS should be using one of these functions in the kernel crypto library:  sha256_transform_ssse3, sha256_transform_avx or sha256_transform_rorx depending on whether the cpu support SSSE3, AVX or AVX2.

**Comment by kpande** [comment]
```
commit 38742209322b5b3d93635a4820b2f9c755aadee8
Author: Tom Caputi <tcaputi@datto.com>
Date:   Thu Mar 15 13:53:58 2018 -0400

    SHA256 QAT acceleration
    
    This patch enables acceleration of SHA256 checksums using Intel
    Quick Assist Technology. This patch also fixes up and refactors
    some of the code from QAT encryption to make the behavior
    consistent.
    
    Reviewed-by: Brian Behlendorf <behlendorf1@llnl.gov>
    Signed-off-by: Chengfeix Zhu <chengfeix.zhu@intel.com>
    Signed-off-by: Weigang Li <weigang.li@intel.com>
    Signed-off-by: Tom Caputi <tcaputi@datto.com>
    Closes #7295
```

this seems to be satisfied since the request was for a quick assist chip to work with ZFS.
